### PR TITLE
Extend the Response with_cors!

### PR DESCRIPTION
**Pull request by ByteAlex**
This PR is a pure QoL PR, which aims to make it easier to configure a CORS policy and apply it to responses using `with_cors` or `with_cors_unsafe` (for convenience).

### CODE REVIEW

**Comment by nilslice** [comment]
Thanks, @ByteAlex - I like the idea of having some utility macros. @zebp - what do you think about adding a new crate for things like this? `worker-ext` or something like that? I'd prefer not to set a precedent of adding all sorts of things to `worker` crate directly, unless they are necessary for _most_ projects. CORS is definitely closer to a main requirement for lots of API-like projects, but I think this could be a nice first for an extension/utility crate.

**Comment by haze** [comment]
:D

**Comment by zebp** [comment]
>  like the idea of having some utility macros. @zebp - what do you think about adding a new crate for things like this? `worker-ext` or something like that?

I think putting utility macros in `worker-macro` isn't a bad idea, but I'm unsure about this PR's `with_cors!`. The type you get is also a `worker::Result<Response>` so I don't know why we should add a macro instead of letting users use `response?.with_cors(whatever)` or `response.and_then(|r| r.with_cors)`.

But other than that macro, I think this PR is pretty good.

File: worker/src/response.rs [hunk 1]
```diff
@@ -193,6 +194,19 @@ impl Response {
...
 
+    /// Sets this response's cors headers from the `Cors` struct.
+    pub fn with_cors(mut self, cors: &Cors) -> Result<Self> {
+        cors.apply_headers(self.headers_mut())?;
+        Ok(self)
+    }
+
+    /// Sets this response's cors headers from the `Cors` struct without checking for the
+    /// `apply_headers` `Result`.
+    pub fn with_cors_unsafe(mut self, cors: &Cors) -> Self {
```
Comment (zebp) [inline worker/src/response.rs:205]: I generally dislike functions that encourage you to not handle your errors and I don't see many scenarios where someone wouldn't care about the cors headers being applied. I think it'd be better if we got rid of this method and users could ignore the result themselves with:
```rust
let cors = whatever;
cors.apply_headers(&mut response).ok();
```

**Review by zebp — CHANGES_REQUESTED** [review]
Looks good for the most part, I have two concerns that I'd like your opinion on.

File: worker/src/cors.rs [hunk 2]
```diff
@@ -0,0 +1,163 @@
...
+/// Example code
+/// ```
+/// pub fn fetch() -> Result<worker::Response, worker::Error> {
+///     let cors = worker::Cors::default();
+///     let response = worker::Response::empty();
+///     return with_cors!(response, cors);
+/// }
+/// ```
+#[macro_export]
+macro_rules! with_cors {
```
[the remark this answers is not in the document]
Comment (ByteAlex) [reply worker/src/cors.rs:17]: True, I totally did not think of this, my mind was stuck in Result#map which was not quite handy.

File: worker/src/response.rs (hunk 1 above)
Comment (ByteAlex) [reply worker/src/response.rs:205]: As above, this was driven by me being stuck with the Result#map approach, fair to remove that too!

**Comment by ByteAlex** [comment]
Though for general knowledge approaching macros:
The worker-marco crate contains only proc-macros so far, if I ever decide to contribute another utility macro-rules macro, it should belong into that crate rather than the respecting file in th worker crate?

**Review by zebp — APPROVED** [review]
Looks good to me,t hanks!

**Pull request merged**